gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -292,4 +306,27 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
       out.endObject();
     }
   }
+
+  /**
+   * Adapter which throws an exception for anonymous and local classes. These types of classes are problematic
+   * because they might capture values of the enclosing context, which prevents proper deserialization and might
+   * also be missing information on serialization since synthetic fields are ignored by Gson.
+   */
+  static class AnonymousLocalClassAdapter<T> extends ReflectiveAdapter<T> {
+    private final Class<?> type;
+
+    AnonymousLocalClassAdapter(Class<?> type) {
+      this.type = type;
+    }
+
+    @Override public void write(JsonWriter out, T value) throws IOException {
+      throw new UnsupportedOperationException("Serialization of anonymous or local class " + type.getName() + " is not supported. "
```

Comment:
Not sure if `UnsupportedOperationException` fits well here. Ideally we should really create a proper exception hierarchy for Gson at some point.

At least the exact exception type is not specified so I guess we can change it later, if necessary.